### PR TITLE
Allow Service connections for push

### PR DESCRIPTION
**Pull request by AlexVTor**
**Task name**: <Name of changed or new pipeline task>
DotNetCoreCLIV2, NpmV1, NuGetCommandV2, UniversalPackagesV0

**Description**: <Describe your changes here>
This PR contains work to allow the use of service connections for the publish scenario for each tasks. 
This is hidden behind a feature flag.
 
**Documentation changes required:** (Y/N) <Please mark if documentation changes are required>
N
**Added unit tests:** (Y/N) <Please mark if unit tests were added or updated according changes>
N
**Attached related issue:** (Y/N) <Please add link to related issue here>

**Checklist**:
- [Y] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [Y] Checked that applied changes work as expected

### CODE REVIEW

File: Tasks/DotNetCoreCLIV2/pushcommand.ts [hunk 1]
```diff
@@ -182,3 +190,57 @@ function dotNetNuGetPushAsync(dotnetPath: string, packageFile: string, feedUri:
...
+            tl.debug("Found external endpoint, will use token for auth");
+            let endpointAuth = tl.getEndpointAuthorization(endpoint, true);
+            let endpointScheme = tl.getEndpointAuthorizationScheme(endpoint, true).toLowerCase();
+            switch(endpointScheme)
+            {
+                case ("token"):
+                    accessToken = endpointAuth.parameters["apitoken"];
+                    break;
+                default:
+                    tl.warning("Invalid authentication type for internal feed. Use token based authentication.");
```
Comment (jmyersmsft) [inline Tasks/DotNetCoreCLIV2/pushcommand.ts:211]: This message is unclear. I'm not sure what I'd do as a customer in response to it.

Also, messages intended for the user need to be localized. Add an entry to 'messages' in task.json and reference it with `tl.loc`.

File: Tasks/NpmV1/npmpublish.ts [hunk 2]
```diff
@@ -47,4 +48,67 @@ export async function getPublishRegistry(packagingLocation: PackagingLocation):
...
+            tl.debug("Found external endpoint, will use token for auth");
+            let endpointAuth = tl.getEndpointAuthorization(endpoint, true);
+            let endpointScheme = tl.getEndpointAuthorizationScheme(endpoint, true).toLowerCase();
+            switch(endpointScheme)
+            {
+                case ("token"):
+                    accessToken = endpointAuth.parameters["apitoken"];
+                    break;
+                default:
+                    tl.warning("Invalid authentication type for internal feed. Use token based authentication.");
```
Comment (jmyersmsft) [inline Tasks/NpmV1/npmpublish.ts:98]: unclear/localize

File: Tasks/NuGetCommandV2/nugetpublisher.ts [hunk 3]
```diff
@@ -405,3 +413,57 @@ function shouldUseVstsNuGetPush(isInternalFeed: boolean, conflictsAllowed: boole
...
+            tl.debug("Found external endpoint, will use token for auth");
+            let endpointAuth = tl.getEndpointAuthorization(endpoint, true);
+            let endpointScheme = tl.getEndpointAuthorizationScheme(endpoint, true).toLowerCase();
+            switch(endpointScheme)
+            {
+                case ("token"):
+                    accessToken = endpointAuth.parameters["apitoken"];
+                    break;
+                default:
+                    tl.warning("Invalid authentication type for internal feed. Use token based authentication.");
```
Comment (jmyersmsft) [inline Tasks/NuGetCommandV2/nugetpublisher.ts:434]: unclear/localize

File: Tasks/UniversalPackagesV0/universalpublish.ts [hunk 4]
```diff
@@ -288,3 +288,42 @@ function authSetup(
         accessToken
     ];
 }
+
+function getAcessToken() : string {
```
Comment (jmyersmsft) [inline Tasks/UniversalPackagesV0/universalpublish.ts:292]: sp: `getAccessToken`

File: Tasks/UniversalPackagesV0/universalpublish.ts [hunk 5]
```diff
@@ -288,3 +288,42 @@ function authSetup(
...
+            tl.debug("Found external endpoint, will try to use token for auth");
+            let endpointAuth = tl.getEndpointAuthorization(endpoint, true);
+            let endpointScheme = tl.getEndpointAuthorizationScheme(endpoint, true).toLowerCase();
+            switch(endpointScheme)
+            {
+                case ("token"):
+                    accessToken = endpointAuth.parameters["apitoken"];
+                    break;
+                default:
+                    tl.warning("Invalid authentication type for internal feed. Use token based authentication.");
```
Comment (jmyersmsft) [inline Tasks/UniversalPackagesV0/universalpublish.ts:309]: unclear/localize

File: Tasks/NpmV1/npmpublish.ts [hunk 6]
```diff
@@ -47,4 +48,67 @@ export async function getPublishRegistry(packagingLocation: PackagingLocation):
             break;
     }
     return npmRegistry;
+}
+
+async function getNpmRegistry(defaultPackagingUri: string, feed: any, authOnly?: boolean, useSession?: boolean) {
```
Comment (jmyersmsft) [inline Tasks/NpmV1/npmpublish.ts:53]: Can you do better than `any` for `feed`'s type?

File: Tasks/DotNetCoreCLIV2/pushcommand.ts (hunk 1 above)
Comment (AlexVTor) [reply Tasks/DotNetCoreCLIV2/pushcommand.ts]: Gotcha, I swapped it out with wording similar to the create service connection prompts:
**The service connection does not use a supported authentication method. Please use a service connection with personal access token based auth**

File: Tasks/NpmV1/npmpublish.ts (hunk 6 above)
Comment (jmyersmsft) [reply Tasks/NpmV1/npmpublish.ts:53]: Looks like the function itself in the library also uses `any`. ☹️ Feel free to dismiss this.

**Review by jmyersmsft — APPROVED** [review]

**Pull request merged**